akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventWriter.scala

```diff
@@ -257,6 +314,98 @@ private[akka] object EventWriter {
+                    currentTransactionId = 1)
+                } else {
+                  sendToJournal(1, Vector(repr))
+                  StateForPid(
+                    Map((repr.sequenceNr, (repr, replyTo))),
+                    currentTransactionId = 1)
+                }
+
+              case Some(state) =>
+                val expectedSeqNr =
+                  if (fillSequenceNumberGaps) state.nextExpectedSeqNr
+                  else repr.sequenceNr
+
+                if (state.seqNrlookupInProgress) {
+                  context.log.trace2(
+                    "Seq nr lookup in progress for persistence id [{}], adding sequence nr [{}] to pending",
+                    persistenceId,
+                    sequenceNumber)
+                  state.copy(waitingForSeqNrLookup = state.waitingForSeqNrLookup :+ ((repr, replyTo)))
+                } else if (fillSequenceNumberGaps && state.latestSeqNr == 0L && sequenceNumber != 1L) {
+                  // we haven't looked up latest yet, and incoming is not the well known 1
+                  askMaxSeqNr(persistenceId, originalErrorDesc = None)
+                  context.log.trace2(
+                    "Seq nr lookup needed for persistence id [{}], adding sequence nr [{}] to pending",
+                    persistenceId,
+                    sequenceNumber)
+                  state.copy(seqNrlookupInProgress = true, waitingForSeqNrLookup = state.waitingForSeqNrLookup :+ ((repr, replyTo)))
+                } else if (sequenceNumber == expectedSeqNr) {
+                  if (state.idle) {
+                    sendToJournal(1, Vector(repr))
+                    state.copy(waitingForReply = Map((repr.sequenceNr, (repr, replyTo))), currentTransactionId = 1)
+                  } else {
+                    // write in progress for pid, add write to batch and perform once current write completes
+                    if (state.waitingForWrite.size >= settings.maxBatchSize) {
+                      replyTo ! StatusReply.error(
+                        s"Max batch reached for pid $persistenceId, at most ${settings.maxBatchSize} writes for " +
+                        "the same pid may be in flight at the same time")
+                      state
+                    } else {
+                      context.log.trace2(
+                        "Writing event in progress for persistence id [{}], adding sequence nr [{}] to batch",
+                        persistenceId,
+                        sequenceNumber)
+                      state.copy(waitingForWrite = state.waitingForWrite :+ ((repr, replyTo)))
+                    }
+                  }
+                } else if (sequenceNumber < expectedSeqNr) {
+                  context.log.trace2("Duplicate seq nr [{}] for persistence id [{}]", sequenceNumber, persistenceId)
+                  replyTo ! StatusReply.success(WriteAck(persistenceId, sequenceNumber)) // duplicate
+                  state
+                } else {
```

Comment:
```suggestion
                } else { // sequenceNumber > expectedSeqNr
```

Maybe also fail if `!fillSequenceNumberGaps` because getting here without that fill enabled would be a bug, right?

Reply:
done